LoginView with RxSwift, RxKeyboard

yooxdb opened the pull request:
Login쪽이 작업량이 많을 것 같아 우선 현재까지 만든 부분에 대한 피드백을 받고자 PR 미리 엽니다.

[ **작업 리스트** ]
- Login
  - [x] Login Page 및 기타 페이지에서 공통적으로 사용할 `RectangularSlimButton` 추가
  - [x] 앱 전반에서 사용할 `UIColor` 추가
  - [x] `RxSwift`를 이용하여 ID, password가 모두 입력되었을 경우에만 로그인 활성화 및 버튼 색상 변경
  - [x] 사용자 계정을 찾을 수 없는 경우 Alert 띄우기
  - [x] ~~ID, password 형식에 대한 추가적인 verification(일정 글자 이상 입력되었을 경우 등등...)~~ 페이스북에 없는 기능이라서 구현X
  - [x] `RxKeyboard`를 이용하여 텍스트필드에 입력 중일 경우 하단의 `"새 계정 생성하기"` 버튼이 올라오도록 설정
- CreateAccount
  - [x] EnterUsernameView + Validation
  - [x] EnterBirthdateView + Validation
  - [x] EnterEmailView + Validation
  - [x] SelectGenderView + Validation
  - [x] EnterPasswordView + Validation

yooxdb committed:
Add LoginPage

- LoginPage추가
- LoginPage에서 사용할 공용 버튼 RectangularSlimButton 추가
- 앱 전반에서 사용할 페이스북 컬러를 enum 형으로 추가

yooxdb committed:
Add verification for LoginPage

- ID와 Password가 모두 입력되어야만 로그인 버튼 활성화 및 버튼 Label 색상 변경
- 우선은 로그인 버튼 활성화와 라벨 색깔 변경에 대한 내용을 분리
- 주석으로 설명 첨부하였습니다. 더 좋은 방법이 있다면 feedback 부탁드립니다.

yooxdb committed:
Install 'RxKeyboard'

- [링크](https://github.com/RxSwiftCommunity/RxKeyboard)

yooxdb committed:
Fix keyboard hiding view

- 가장 하단의 "새 계정 생성하기" 버튼이 위의 textfield가 입력 중일 때는 함께 위로 올라오도록 설정
- NotificationCenter 없이 Rx로 최대한 구현해보려 했는데 view의 layout을 각 ~view.swift 파일에서 정하고 있다보니 가장 아래 버튼의 bottomConstraint를 별도 변수로 선언 뒤, ViewController에서 조절하는 방식정도가 최선이었습니다. 어딘가 깔끔하지 않아서 좀 애매하네요.

shp7724 commented on Facebook/Facebook/Controller/LoginViewController.swift:
```diff
@@ -0,0 +1,101 @@
+            }).disposed(by: disposeBag)
+
+        // Keyboard의 높이에 따라 "새 계정 만들기" 버튼 위치 조정
+        RxKeyboard.instance.visibleHeight
+            .drive(onNext: { [weak self] keyboardVisibleHeight in
+                guard let self = self else { return }
+                guard let bottomConstraint = self.loginView.bottomConstraint else { return }
+
+                if keyboardVisibleHeight == 0 {
+                    bottomConstraint.constant = -16.0
+                } else {
+                    let height = keyboardVisibleHeight - self.view.safeAreaInsets.bottom
+                    print(height)
+                    bottomConstraint.constant = -height - 16.0
+                }
+                self.loginView.layoutIfNeeded()
+            })
+            .disposed(by: disposeBag)
```
아이디 필드와 패스워드 필드를 번갈아가면서 탭하면 `새 계정 만들기` 버튼의 애니메이션에 글리치가 생기네요.

임시방편으로 `.debounce(.milliseconds(100))`처럼 디바운스를 걸어주면 해결이 되긴 하는데, 애초에 `keyboardVisibleHeight`이 순간적으로 0으로 바뀌는 원인을 찾아서 해결하는게 더 바람직한 방법일 것 같습니다. (왜 그런지는 저도 모르겠네요)

---

디바운스를 걸어주니 이젠 애니메이션 적용이 안되는군요 ㅠ
조금 더 생각을 해봐야 할 것 같습니다.

yooxdb replied:
말씀해주신 부분 확인했습니다! 필드 두 개를 연달아 바로 눌러볼 생각은 못했네요ㅎㅎ..저도 해결 방법 조금 더 고민해보도록 하겠습니다

--------
[RxKeyboard Issue 링크](https://github.com/RxSwiftCommunity/RxKeyboard/issues/105) 별도의 설정이 없으면 계속해서 변하는 Keyboard의 높이가 방출되지 않고 시작과 끝만 방출되는 것 같던데 링크에 달린 답변대로 진행하기에는 지금으로써는 굳이..? 싶어서 패스했습니다

shp7724 replied:
원인을 찾은 것 같습니다.

```swift
view.rx.tapGesture()
            .asObservable()
            .subscribe(onNext: { [weak self] _ in
                guard let self = self else { return }
                if self.loginView.idTextField.isEditing {
                    self.loginView.idTextField.endEditing(true)
                } else if self.loginView.passwordTextField.isEditing {
                    self.loginView.passwordTextField.endEditing(true)
                }
            })
            .disposed(by: disposeBag)
```

"아무 곳(`view`)"에 텍스트 필드도 포함되는 바람에, 다른 텍스트 필드로 이동할 때 endEditing이 한번 호출되고(높이->0) 다시 first responder가 되면서(높이->300 얼마) 글리치가 생기는 것 같습니다.

shp7724 commented on Facebook/Facebook/Controller/LoginViewController.swift:
```diff
@@ -0,0 +1,101 @@
+                guard let bottomConstraint = self.loginView.bottomConstraint else { return }
+
+                if keyboardVisibleHeight == 0 {
+                    bottomConstraint.constant = -16.0
```
그냥 질문인데 여기서 -16이 어떤 값을 의미하는건가요?

yooxdb replied:
View의 bottom으로부터 -16.0만큼 떨어지는 것이니 `createAccountButton`이 아래 마진에서 16.0만큼 위에 위치하게 됩니다

shp7724 replied:
아 그냥 임의로 설정하신 값이었군요 이해했습니다!

shp7724 commented on Facebook/Facebook/Controller/LoginViewController.swift:
```diff
@@ -0,0 +1,101 @@
+        hasEnteredBoth
+            .subscribe(onNext: { [weak self] result in
+                guard let self = self else { return }
+                self.loginView.loginButton.changeLabelTextColor(to: {
+                    return result ? UIColor.white : UIColor.systemGray3
+                }())
```
(큰 영향은 없겠지만) UI 업데이트하는 코드는 메인 스레드에서 실행하는게 원칙이므로, `.observeOn(MainScheduler.instance)`를 추가해주는게 좋은 방식일 것 같다는 생각이 들었습니다. 어떻게 생각하시나요?

yooxdb replied:
넵 `.observe(on: MainScheduler.instance)` 추가했습니다!

shp7724 replied:
위에 RxKeyboard에 `.drive`를 쓰셨길래 검색해보니 이거랑 같은 기능을 하는 것 같네요. 하나 배워갑니다..!

shp7724 reviewed (commented):
디테일한 부분까지 잘 구현해주신 것 같습니다..!
RxSwift 사용 경험이 많지 않아서 리뷰하기가 조금 조심스럽지만 몇 가지 코멘트 남겼으니 확인바랍니다 :)

shp7724 commented on Facebook/Facebook/Controller/LoginViewController.swift:
```diff
@@ -0,0 +1,101 @@
+        hasEnteredBoth
+            .subscribe(onNext: { [weak self] result in
+                guard let self = self else { return }
+                self.loginView.loginButton.changeLabelTextColor(to: {
```
이 부분은 그냥 이렇게 써도 되지 않나요?

```swift
self.loginView.loginButton.changeLabelTextColor(to: result ? UIColor.white : UIColor.systemGray3)
```

yooxdb replied:
넵 그렇게 써도 상관 없습니다. 원래 중간에 다른 코드를 넣느라 closure형식으로 작성했는데 수정 과정에서 중간의 내용만 빠지고 틀은 그대로 남다보니 지금 상태네요

KWSMooBang reviewed (commented):
수고하셨습니다! 
신홍님이 말씀하신 부분 외에는 문제 없는 것 같습니다.

shp7724 committed:
애니메이션 버그 수정

yooxdb committed:
Merge pull request #11 from wafflestudio19-5/fix/auth-ui-glitch

로그인 페이지 애니메이션 버그 수정

yooxdb committed:
Add BaseAlertController + BaseSignUpView

- alert를 띄울 때 사용할 extension 추가
- 가입 절차에서 공통적으로 보이는 view에 대한 BaseSignUpView 추가

yooxdb committed:
Merge remote-tracking branch 'origin/feat/authentication' into feat/authentication

yooxdb committed:
Chore: RootTabBarController

- viewWillAppear -> setTabBarController로 수정

yooxdb committed:
Minor Fix: FacebookTextField feat

- 기존: 생성 시, CGFloat형 width를 받아서 widthAnchor 설정
- 수정: width를 따로 넘겨받지 않고, FacebookTextField를 가지고 있는 view에서 constraint 지정을 통해 설정되도록

yooxdb committed:
Add EnterUsernameView

- SignUp의 첫 단계인 이름 입력 View 및 ViewController 추가

yooxdb committed:
Minor Fix: RectangularSlimButton feat

- 기존: 생성 시, CGFloat형 width를 받아서 widthAnchor 설정
- 수정: width를 따로 넘겨받지 않고, RectangularSlimButton을 가지고 있는 view에서 constraint 지정을 통해 설정되도록

yooxdb committed:
Chore: Fix typo

- LoginView의 LoginButton, CreateAccountButton constraint에서의 오타 수정

yooxdb committed:
Add EnterBirthdateView

- EnterBirthdateViewController도 함께 추가
- Facebook 가입 절차 중 두 번째인 생년월일 입력 페이지
- UIDatePicker 이용하여 구현

shp7724 commented on Facebook/Facebook/View/SignUpViewGroup/EnterBirthdateView.swift:
```diff
@@ -0,0 +1,89 @@
+    }
+    
+    // 페이스북에서는 생년월일을 1905년부터 설정 가능합니다
+    private func setLimitForBirthdatePicker() {
+        var minimumYear = DateComponents()
+        minimumYear.year = 1905
+        birthDatePicker.minimumDate = Calendar.autoupdatingCurrent.date(from: minimumYear)
+    }
+    
```
현재 날짜 이후로 설정할 수 없게끔 추가해주면 더 좋을 것 같습니다.

yooxdb replied:
네 관련 내용 추가하겠습니다:) 그리고 페이스북이 만 14세 미만일 경우에는 계정을 생성할 수 없도록 막아두었는데 이거 관련 validation은 추후 추가하겠습니다

shp7724 commented on Facebook/Facebook/Components/BaseAlertController.swift:
```diff
@@ -0,0 +1,16 @@
+import UIKit
+
+extension UIViewController {
+    func alert(title: String, message: String, action: String) {
```
`extension`이 Components 폴더에 들어가있는게 좀 어색한 느낌인데 따로 Extensions 폴더를 만들어서 거기로 빼는게 어떤가요?

shp7724 commented on Facebook/Facebook/Controller/LoginViewController.swift:
```diff
@@ -0,0 +1,116 @@
+            print("createAccountButton tapped")
+            guard let self = self else { return }
+            let viewController = EnterUsernameViewController()
+            self.navigationController?.pushViewController(viewController, animated: true)
```
이 버튼 클릭했을때 Push 애니메이션이 작동하지 않는데 유림님은 되시나요? (이름 입력하고 다음 넘어가는 버튼 클릭해도 동일합니다)

yooxdb replied:
실제 페이스북 가입 절차에서는 createAccountButton을 누른 이후로 Pop/push 애니메이션이 없어서 일부러 애니메이션을 제거해둔 상태인데(`EnterBirthdateViewController` line 34~) 여기서 `UIView.setAnimationEnabled(false)`로 해두어서 상관이 없는 다른 View까지 적용되는 것 같습니다. 이 부분은 수정할게요

shp7724 commented on Facebook/Facebook/Controller/LoginViewController.swift:
```diff
@@ -0,0 +1,116 @@
+            print("login button tapped")
+            
+            // 로그인에 실패한 경우 보이는 알림창 sample
+            self.alert(title: "잘못된 사용자 이름", message: "입력하신 사용자 이름에 해당하는 계정을 찾을 수 없습니다. 사용자 이름을 확인하고 다시 시도하세요.", action: "확인")
```
이것도 애니메이션이 작동 안하는 것 같습니다. 🤔

yooxdb replied:
위의 내용에 답변 남겼습니다!

shp7724 reviewed (commented):
작업을 엄청 많이 해주셨네요. 👍🏻 

---


프로젝트가 너무 커지기 전에 한번 정리하는게 좋을 것 같은데요.

지금은 Model, View, Controller 각 폴더 내에 기능별로 그룹을 만들어 둔 형태인데, 
제 생각엔 각 기능을 모듈화해서 그룹을 생성하고, 각 모듈 내에 MVC 폴더가 존재하는 형태로 하는게 협업할 때 조금 더 깔끔한 구조가 되지 않을까 싶습니다. (저희 백엔드도 이런 구조를 따르는 것으로 알고 있습니다)

(그리고 사소한 거긴 한데 Model/View/Controller 폴더명에 s를 붙이는게 의미/통일성 측면에서 좋지 않을까요?)

@KWSMooBang 이거 관련해서 의견 나눠보면 좋을 것 같아요!

KWSMooBang commented on Facebook/Facebook/Controller/SignUpViewControllerGroup/EnterUsernameViewController.swift:
```diff
@@ -0,0 +1,38 @@
+        bindView()
+    }
+    
+    private func bindView() {
+        enterUsernameView.nextButton.rx.tap.bind {
+            let enterBirthdateViewController = EnterBirthdateViewController()
+            self.navigationController?.pushViewController(enterBirthdateViewController, animated: false)
+        }.disposed(by: disposeBag)
+    }
```
`nextButton`을 눌렀을 때 `textField`가 비어있어도 `EnterBirthdateView`로 넘어가 지는 것 같습니다. 
`textField`가 비어있는 상태에서 `nextButton`을 눌렀을 때는 다음 뷰로 넘어가지 않고, 빨간 경고 메세지를 띄우는 것 같은 기능이 필요할 것 같습니다.

yooxdb replied:
넵 일단은 view만 먼저 짜두고 아직 validation 관련한 기능은 넣어놓지 않았습니다. 추후 추가하겠습니다

KWSMooBang reviewed (commented):
`LoginView`에서 `textField`를 선택해 키보드가 올라와 있는 상태에서 계정 생성 버튼을 누르면, 다시 `LoginView`로 돌아왔을 때 기존에 `textField`에 입력된 입력값이 남아있고, 키보드가 올라와 있는 상태가 되는 것 같습니다. 다시 돌아왔을 때 이것들을 초기화 해주는 것이 어떨까요?

yooxdb committed:
Fix DatePicker for Birthdate

- maximumDate를 현재 날짜로 설정(오늘까지만 설정 가능, 그 이후는 불가능)
- todo: 만 14세 미만 검사 추가

yooxdb committed:
Fix animation for SignUp

- push/popViewController 애니메이션 모두 활성화하는 것으로 수정

yooxdb committed:
Add EnterEmailView

- validation은 추후 넣을 예정입니다

yooxdb committed:
Refactor: add BaseSignUpViewController

- 회원가입 과정에서 사용되는 viewController이 BaseSignUpViewController를 상속하도록 구조 변경

yooxdb committed:
Add SelectGenderView

- 성별을 선택하는 SelectGenderView + ViewController 추가
- 성별을 선택하는 리스트는 UITableView를 이용하여 구현(단, 스크롤 없이 내부의 contents를 한번에 다 보여줘야하기 때문에 UITableView를 상속받는 custom class인 ContentSizeFitTableView 추가)
- todo: validation, 선택한 Row의 정보 가공(필요할 시)

yooxdb committed:
Add EnterPasswordView

- 회원가입 마지막 절차인 EnterPasswordView + ViewController 추가
- 실제 앱에는 약관 동의 등이 있는데 이 부분은 빼고 구현했습니다

yooxdb committed:
Add Username Validation

- 하단의 경고 label을 띄우기 위해 LabelWithPadding 추가
- 버튼이 tap될 때에만 경고 메시지가 변하도록 작성
- 두 칸이 모두 비어있을 때 / 성만 입력되었을 때 / 이름만 입력되었을 때에 따라 경고 메시지가 다르게 보이도록 작성

yooxdb commented on Facebook/Facebook/View/SignUpViewGroup/EnterUsernameView.swift:
```diff
@@ -30,24 +33,44 @@ class EnterUsernameView: UIView {
         fatalError("init(coder:) has not been implemented")
     }
     
+    func setAlertLabelText(as text: String) {
```
경고 label은 view에서 자리가 고정되어 있는 상태에서 alpha값이 변하는 방식이 아니라 view에 추가/삭제되는 방식이라는 점 + 하단의 nextButton의 constraint 지정을 위해 textfield와 alertLabel을 하나의 UIStackView로 묶는 방식을 사용하였습니다.

yooxdb committed:
Refactor: add Extensions

- Extensions를 따로 빼고, 간결한 pushViewController 사용을 위한 push 추가

yooxdb committed:
Add Birthdate Validation

- 생일을 기준으로 나이를 계산하여, 5세 이상이면 다음 페이지로 넘어가고 5세 미만인 경우에는 alertLabel을 띄우며 다음 페이지로 넘어가지 않는 점 반영
- alertLabel을 띄우는 방식은 username validation 때와 동일한 방식을 사용
- ViewController 쪽의 주석 참고 부탁드립니다

KWSMooBang commented on Facebook/Facebook/Controller/SignUpViewControllerGroup/EnterEmailViewController.swift:
```diff
@@ -0,0 +1,37 @@
+            }
+        }.disposed(by: disposeBag)
+        
+        customView.nextButton.rx.tap.bind { [weak self] _ in
+            guard let self = self else { return }
+            self.push(viewController: SelectGenderViewController())
+        }.disposed(by: disposeBag)
```
이 부분에서 이메일을 입력하지 않아도 넘어가 지는 것 같습니다. 이름 입력 화면과 동일하게 메세지를 띄워주는 게 좋을 것 같아요!

yooxdb replied:
슬랙에 말씀드렸다시피 아직 Validation은 Username이랑 Birthdate에만 적용되어있습니다! 이메일/성별/비밀번호 Validation 작업 후에 최종 리뷰 부탁드릴 예정입니다 한번에 모두 리뷰하시려면 양이 부담스러우실까봐 중간에 리뷰 요청드렸습니다:)

KWSMooBang commented on Facebook/Facebook/Controller/SignUpViewControllerGroup/EnterPasswordViewController.swift:
```diff
@@ -0,0 +1,38 @@
+            }
+        }.disposed(by: disposeBag)
+        
+        customView.nextButton.rx.tap.bind {
+            let rootTabBarController = RootTabBarController()
+            guard let sceneDelegate = UIApplication.shared.connectedScenes.first?.delegate as? SceneDelegate else { return }
```
이 부분도 동일하게 입력하지 않아도 넘어갈 수 있는 것 같습니다.

KWSMooBang commented on Facebook/Facebook/Controller/SignUpViewControllerGroup/SelectGenderViewController.swift:
```diff
@@ -0,0 +1,52 @@
+                cell.isSelected = false
+            }.disposed(by: disposeBag)
+        
+        customView.nextButton.rx.tap.bind { [weak self] _ in
+            guard let self = self else { return }
+            self.push(viewController: EnterPasswordViewController())
+        }.disposed(by: disposeBag)
```
이 부분도 선택 없이 넘어갈 수 있는 것 같습니다.

KWSMooBang reviewed (commented):
수고하셨습니다! 
입력 값 없이 다음화면으로 넘어갈 수 있는 것만 수정하면 괜찮을 것 같습니다.

shp7724 commented:
수고하셨습니다!
비밀번호 텍스트 필드의 내용이 마스킹 없이 보이는 점만 제외하면 다 괜찮은 것 같습니다.

yooxdb committed:
Add Email Validation

- 페이스북 앱에서는 textField가 빈칸만 아니면 되지만 클론 앱에서는 정규식을 이용한 간단한 validation을 추가하였습니다
- "영어 대문자, 소문자 혹은 숫자 1개 이상"@"영어 소문자 1개 이상"."영어 소문자 2개 이상" 일 경우 valid하다고 판단 후 다음 페이지로 이동
- 예시: (valid) abc@a.kr 혹은 8312@abc.com
           (invalid) abc@a. 혹은 a90@naver.cOM

yooxdb committed:
Add Gender Validation

- 성별 선택 페이지는 다른 페이지와 다르게 alertMessageLabel이 상단에 뜨는 점 반영
- 아무것도 선택하지 않고 "다음" 버튼을 누를 시 "여성", "남성" label 색상이 빨간색으로 뜨는 점 반영

yooxdb committed:
fix typo

yooxdb committed:
Add Password Validation

- 페이스북 앱에서는 password 입력하는 textField가 secure하게 설정되어 있지 않더군요..클론 앱에서는 secureTextField로 설정하였습니다
- 비밀번호 칸이 비어있을 때, 1자 이상 6자 미만일 때, 6자 이상이지만 부적절한 문자가 표현되어 있을 때, 6자 이상이고 적절한 문자로만 구성되어 있을 때(총 4가지) 상황에 따른 alertLabelMessage 설정하였습니다

yooxdb committed:
Refactor: SignUpViews extend BaseSignUpView

- 기존: 각 SignUpView에서 BaseSignUpView를 변수로 지니고 있고, 각 페이지의 용도에 맞는 alertLabel와 다른 UIView를 지니고 있음
- 수정 후: alertLabel과 verticalStackWithAlertLabel을 BaseSignUpView가 지니게 하고, 다른 SignUpView들이 BaseSignUpView를 상속하도록 수정

yooxdb committed:
Refactor: Add BehaviorRelay for LoginViewController

- id와 password 입력값을 BehaviorRelay로 받도록 수정

shp7724 commented on Facebook/Facebook/Controller/SignUpViewControllerGroup/EnterEmailViewController.swift:
```diff
@@ -29,9 +64,30 @@ class EnterEmailViewController: BaseSignUpViewController<EnterEmailView> {
+extension EnterEmailViewController {
+    private func isValid(email: String) -> Bool {
+        // "영어 대문자, 소문자 혹은 숫자 1개 이상"@"영어 소문자 1개 이상"."영어 소문자 2개 이상"
+        let pattern = "^[A-Za-z0-9]+@[a-z]+\\.[a-z]{2,}$"
```
정규식이 `username@do-main.co.kr` 이런 케이스를 처리해주지 못하는 것 같습니다!

yooxdb replied:
`username@naver.com`같은 케이스만 생각하고 만들었는데 `@` 뒤에 .이 2개 오는 경우도 포함하도록 수정해보겠습니다! 사실 페이스북에서는 빈칸만 아니면 다 통과를 시켜서ㅎㅎ저희가 추가하기 나름일 것 같네요.

shp7724 replied:
제가 방금 해보니 페이스북은

1. 일단 검증 없이 각 step 무조건 통과
2. 마지막 단계에서 "다음" 버튼을 누르면 입력된 모든 정보 백엔드에 전송
3. 백엔드에서 Validation 수행한 뒤 콜백을 받아서 검증이 필요한 step을 전부 재작성하도록 유도하네요.

그래서 프론트에서는 1차로 간단하게 거르고,
2차는 백엔드 내장 validator(로직이 상당히 복잡해서 프론트에서 그대로 구현하기에는 무리가 있을 것 같습니다.)로 검증하는게 좋을 것 같습니다.


## 참고 영상
https://user-images.githubusercontent.com/33917774/146679959-56515589-da8e-45ac-b6e3-cd7eb139b732.MP4

yooxdb replied:
제가 회원가입 절차를 거쳤을 때에는 백엔드 내장 validator을 통과하지 못하는 field가 있을 경우 "계정을 생성할 수 없습니다."라고 뜨며 계정 생성이 일정 시간 안되는 현상만 반복되었는데 이전 칸으로 돌아가는 모습은 처음 보네요. 백엔드 내장 validator을 통과하지 못한 경우 해당 페이지로 pop하는 기능은 저희가 페이스북 클론 앱을 통해 구현해야할 스펙 중 priority가 높아보이진 않아서 이 기능은 issue로 남겨놓고 나중에 돌아와서 작업하는 쪽이 나을 것 같은데, 혹시 신홍님께서는 어떻게 생각하시나요?

shp7724 replied:
네 제 생각엔 여러 validator 중 가장 복잡도가 높은게 PasswordValidator여서요,
username이나 email 같은 건 프론트에서 100% 거를 수 있다고 가정하고,
마지막 password 입력 단계에서 *다음* 버튼을 누르면 지금까지 저장된 정보를 백엔드에 보낸 뒤, (← 어차피 해야하는 작업이죠)
회원가입 성공 응답을 받으면 그대로 진행하고,
실패 응답을 받으면 Response에서 에러 메시지를 뽑아서 password 입력필드 밑에 빨간색으로 넣는 방식으로 하면 충분할 것 같다고 생각했습니다.

이정도면 원래 해야하는 작업에서 복잡도를 크게 높이지 않는 선에서 가능할 것 같아요.

yooxdb replied:
실패 응답을 받으면 Response에서 에러 메시지를 뽑아서 password 입력필드 밑에 빨간색으로 넣자는 말씀은 백엔드에서의 validator은 password에 대해서만 진행하고, 에러가 날 경우 password에 대한 에러 메시지만 띄우자는 말씀이 맞으실까요?

shp7724 replied:
백엔드 Validator는 모든 필드(이메일, 패스워드 ..)에 적용되는걸로 알고 있습니다. 다만 이메일 등에 대한 백엔드 Validator는 반드시 통과한다고 가정하고 드린 말씀이었어요.

실제 백엔드에서 가입 실패시 Response가 어떻게 나올지는 모르겠지만,

```json
{
    "success": false,
    "message": "비밀번호는 특수문자가 포함되어야 합니다."
}

{
    "success": false,
    "message": "유효한 이메일이 아닙니다."
}
```

만약 이런식으로 뜬다고 가정하면 Password 입력 페이지에서 `message` 필드를 경고 문구로 띄우자는 뜻이었어요.
물론 두번째처럼 이메일에서 틀리면 조금 이상하게 보이긴 하겠지만, 이메일은 프론트에서 최대한 거른다고 가정하면 확률은 높지 않을거라고 생각했습니다. (토이플젝인것도 감안해서요.)

shp7724 commented on Facebook/Facebook/Controller/SignUpViewControllerGroup/EnterPasswordViewController.swift:
```diff
@@ -29,10 +61,36 @@ class EnterPasswordViewController: BaseSignUpViewController<EnterPasswordView> {
+        if password.isEmpty { return PasswordValidation.emptyLetters }
+        else if password.count < 6 { return PasswordValidation.lessThanSixLetters }
+        else {
+            let pattern = "^[A-Za-z0-9!&]*$"
```
> 숫자, 영문, 특수기호(!, & 등)를 조합한 여섯 자리 이상의 비밀번호를 입력하세요.

정규식과 위 규칙이 좀 다르게 작동할 것 같습니다.
(!, & 이외에 다른 특수문자가 포함되면 invalid, 특수문자가 전혀 없어도 valid)

별개로 백엔드에도 기본적으로 내장된 PasswordValidator 로직이 있을거라서(숫자로만 이루어진 문자, 비슷한 문자 등),
프론트에서 1차로 거르고 최종 유효성은 백엔드에서 검증받는 방식으로 하는 게 좋을 것 같습니다.

yooxdb replied:
네 실제 페이스북에서도 특수문자가 전혀 없어도 valid 합니다! 그리고 !와 & 이외의 특수기호가 들어가면 invalid하다고 뜹니다
위 규칙은 Facebook의 글귀를 그대로 가져다 놓은 것인데 번역체라 그런건지 말이 조금 모호하더군요...

shp7724 commented on Facebook/Facebook/Controller/SignUpViewControllerGroup/EnterPasswordViewController.swift:
```diff
@@ -29,10 +61,36 @@ class EnterPasswordViewController: BaseSignUpViewController<EnterPasswordView> {
+                self.customView.setAlertLabelText(as: isValidPassword.message())
+                
+                if isValidPassword == .valid {
+                    // TODO: User 등록
```
이 시점에서 다른 ViewController에서 저장된 정보를 어떻게 가져올 예정이신가요?

yooxdb replied:
신홍님 작업하신 코드를 보니 User 모델을 추가하셨길래 일단 현재 상태에서 머지를 진행하고, 신홍님 작업하던 브랜치를 기반으로 새 브랜치를 파서 가입의 첫 단계인 `EnterUserName`에서 `User`형 객체를 하나 선언한 뒤, 각 페이지에서 User의 속성을 하나씩 추가하여 (ex: `user.email = emailAddress.value`) 마지막 단계에서 완성된 User 객체를 서버에 등록시키는 방법으로 하고자 했는데, 혹시 제가 잘못 생각하고 있는 부분이 있을까요..?

shp7724 replied:
네 전체적인 흐름은 저랑 같은 생각이신 것 같습니다.
다만 제 생각엔 `User`를 그대로 사용하기보다는 회원가입 목적에 맞게 모델을 따로 생성하는 게 나을 것 같습니다.
`password` 같은 필드는 회원가입할때만 필요한데 그걸 서비스 전역에서 사용할 `User` 객체에 추가하는 것은 적절하지 않을 것 같아서요.

[백엔드](http://3.34.188.255/swagger/)에서는 동일한 목적으로 `UserCreate`라는 별도의 모델을 사용하고 있는 것 같네요.

yooxdb replied:
아 넵넵 백엔드에서는 `UserCreate`라는 별도 모델을 사용하고 있군요. 그렇다면 iOS에서도 회원 가입을 위한 Model을 별도로 생성해서 작업하면 될 것 같습니다. 이 부분은 위에서 말씀드렸다시피 이 브랜치 머지 이후에 제가 새로 브랜치 파서 작업해보겠습니다!

shp7724 reviewed (approved):
고생하셨습니다! 
이제 슬슬 머지해도 될 것 같은데 아직 남은 작업이 있으신가요?

yooxdb committed:
Fix Email validator regex

- user_name@domain-name.ac.kr 과 같은 이메일도 validate하도록 수정
- regex: 영어 대문자, 소문자, 숫자, 혹은 -, _ 1개 이상"@"영어 소문자 1개 이상"."영어 소문자 혹은 ., -을 포함하여 2개 이상

KWSMooBang reviewed (commented):
수고하셨습니다!
이대로 merge해도 될 것 같습니다.
